performance_manager/lib/l1_full_trip_events.py

```diff
@@ -0,0 +1,223 @@
+        ).where(
+            (VehiclePositionEvents.is_moving == sa.true())
+            & (
+                VehiclePositionEvents.pk_id.notin_(
+                    sa.select(FullTripEvents.fk_vp_moving_event)
+                )
+            )
+            & (VehiclePositionEvents.hash.notin_(dupe_hash_query))
+        ),
+        fk_rename_field="fk_vp_moving_event",
+    )
+
+    stopped_vp_events = EventsToMerge(
+        select_query=sa.select(
+            VehiclePositionEvents.pk_id,
+            VehiclePositionEvents.stop_sequence,
+            VehiclePositionEvents.stop_id,
+            VehiclePositionEvents.direction_id,
+            VehiclePositionEvents.route_id,
+            VehiclePositionEvents.start_date,
+            VehiclePositionEvents.start_time,
+            VehiclePositionEvents.vehicle_id,
+        ).where(
+            (VehiclePositionEvents.is_moving == sa.false())
+            & (
+                VehiclePositionEvents.pk_id.notin_(
+                    sa.select(FullTripEvents.fk_vp_stopped_event)
+                )
+            )
+            & (VehiclePositionEvents.hash.notin_(dupe_hash_query))
+        ),
+        fk_rename_field="fk_vp_stopped_event",
+    )
+
+    stopped_tu_events = EventsToMerge(
+        select_query=sa.select(
+            TripUpdateEvents.pk_id,
+            TripUpdateEvents.stop_sequence,
+            TripUpdateEvents.stop_id,
+            TripUpdateEvents.direction_id,
+            TripUpdateEvents.route_id,
+            TripUpdateEvents.start_date,
+            TripUpdateEvents.start_time,
+            TripUpdateEvents.vehicle_id,
+        ).where(
+            (
+                TripUpdateEvents.pk_id.notin_(
+                    sa.select(FullTripEvents.fk_tu_stopped_event)
+                )
+            )
+        ),
+        fk_rename_field="fk_tu_stopped_event",
+    )
+
+    return {
+        "moving_vp": moving_vp_events,
+        "stopped_vp": stopped_vp_events,
+        "stopped_tu": stopped_tu_events,
+    }
+
+
+def load_and_transform(
+    events_list: Dict[str, EventsToMerge], db_manager: DatabaseManager
+) -> None:
+    """
+    load and transform each set of events from RDS L0 tables
+    """
+    expected_hash_columns = [
+        "stop_sequence",
+        "stop_id",
+        "direction_id",
+        "route_id",
+        "start_date",
+        "start_time",
+        "vehicle_id",
+    ]
+
+    for event_type in events_list.values():
+        event_type.merge_dataframe = db_manager.select_as_dataframe(
+            event_type.select_query
+        )
+
+        if event_type.merge_dataframe.shape[0] == 0:
+            event_type.merge_dataframe = pandas.DataFrame(
+                columns=["pk_id", "hash"]
+            )
+        else:
+            event_type.merge_dataframe = add_event_hash_column(
+                event_type.merge_dataframe,
+                expected_hash_columns=expected_hash_columns,
+            )[["pk_id", "hash"]]
```

Comment:
add `hash` column based on category column, `hash` will be used as primary key in `FullTripEvents` table

Reply:
why do we have the `shape[0] == 0` on the if gate here?

Reply:
That's a check on how many rows are in the dataframe. If no rows are returned it just creates an empty dataframe with the required columns that can pass through the merge logic. 

I'll add a comment here explaining this also